selftest/cgroup/main.go

```diff
@@ -0,0 +1,81 @@
+package main
+
+import "C"
+
+import (
+	"encoding/binary"
+	"os"
+	"os/exec"
+
+	"fmt"
+
+	bpf "github.com/aquasecurity/libbpfgo"
+)
+
+func main() {
+	bpfModule, err := bpf.NewModuleFromFile("main.bpf.o")
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+	defer bpfModule.Close()
+
+	err = bpfModule.BPFLoadObject()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(-1)
+	}
+
+	cgroupRootDir := "/sys/fs/cgroup/unified"
+	if _, err := os.Stat(cgroupRootDir); os.IsNotExist(err) {
+		cgroupRootDir = "/sys/fs/cgroup"
```

Comment:
We have to check whether this is a cgroupv2 directory or not (cgroup v1 is not supported). Best way to do this is to check for the "release_agent" file existence iirc (cgroup v2 doesn't have it).

Reply:
https://man7.org/linux/man-pages/man7/cgroups.7.html

```
Note that on many modern systems, [systemd(1)](https://man7.org/linux/man-pages/man1/systemd.1.html) automatically mounts
       the cgroup2 filesystem at /sys/fs/cgroup/unified during the boot
       process.
```

This is not so true, since in my arch based it's mounted at /sys/fs/cgroup. So checking whether it's unified or not would not help in this case.

```bash
$ mount | grep cgroup
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime)

$ mount | grep cgroup | cut -d " " -f 1
cgroup2
```

Perhaps checking whether `/sys/fs/cgroup/cgroup.controllers` is available then it would be v2?

```
The new behaviors in cgroups v2 are summarized here, and in some
       cases elaborated in the following subsections.

       1. Cgroups v2 provides a unified hierarchy against which all
          controllers are mounted.
...
       3. Active cgroups must be specified via the files
          cgroup.controllers and cgroup.subtree_control.
```

Or checking as systemd does: https://github.com/systemd/systemd/blob/5c19169fe95f436a233f92fc62fdca783a0fc5ec/src/basic/cgroup-util.c#L2099

It's summarized here: https://unix.stackexchange.com/a/617824/47040

Reply:
@rafaeldtinoco what main distros do we need to keep in mind?

Reply:
The thing is cgroupv2 support might be available even if cgroupv2 is not mounted. Distros might choose to mount cgroupv1, cgroupv1+v2 or cgroupv2 only. The way we solved that in tracee is to temporarily mount a cgroupv2 filesystem if not mounted, check availability and use that directory if needed.

Systemd provides a way to choose if you want to have mixed cgroup v1 and cgroup v2 mounted filesystems (systemd.unified_cgroup_hierarchy=XXX). For this test I would say checking if cgroupv2 is mounted (through /proc/mounts and check "cgroup2") would be enough (and to use that dir).

Reply:
fixed

Reply:
@mozillazg Thank you for the time into this.